src/Grapevine.Tests/Server/PublicFolderFacts.cs

```diff
@@ -18,133 +19,209 @@ private static string GenerateUniqueString()
+                folder.DirectoryListing.Count(x => x.Value == Path.Combine(folder.FolderPath, defaultFileName2)).ShouldBe(2);
+
+                folder.CleanUp();
             }
+        }
 
+        public class FileSystemWatcherProperty
+        {
             [Fact]
-            public void ReturnsNullWhenDefaultFileDoesNotExist()
+            public void CanProvideCustomWatcher()
             {
-                var folder = GenerateUniqueString();
-                var root = new PublicFolder();
+                var folder = new PublicFolder();
+                var watcher = new FileSystemWatcher
+                {
+                    Path = folder.FolderPath,
+                    Filter = "*.jpg",
+                    EnableRaisingEvents = true,
+                    IncludeSubdirectories = true,
+                    NotifyFilter = NotifyFilters.FileName
+                };
 
-                var folderpath = Directory.CreateDirectory(Path.Combine(root.FolderPath, folder)).FullName;
-                if (!Directory.Exists(folderpath)) throw new Exception("Folder to test did not get created");
+                folder.Watcher.Filter.ShouldNotBe(watcher.Filter);
 
-                root.FilePathGetter($"/{folder}").ShouldBeNull();
+                folder.Watcher = watcher;
 
-                CleanUp(folderpath);
+                folder.Watcher.Filter.ShouldBe(watcher.Filter);
             }
 
             [Fact]
-            public void ReturnsExistingFilePath()
+            public void DoesNotUpdateToNull()
             {
-                var root = new PublicFolder();
-                var file = GenerateUniqueString();
+                var folder = new PublicFolder();
+                var watcher = folder.Watcher;
 
-                var filepath = Path.Combine(root.FolderPath, file);
-                using (var sw = File.CreateText(filepath)) { sw.WriteLine("Hello"); }
+                folder.Watcher = null;
 
-                root.FilePathGetter(file).ShouldBe(filepath);
-                root.FilePathGetter($"/{file}").ShouldBe(filepath);
-
-                File.Delete(filepath);
+                folder.Watcher.ShouldNotBeNull();
+                folder.Watcher.Equals(watcher).ShouldBeTrue();
             }
 
             [Fact]
-            public void ReturnsExistingDefaultFilePath()
+            public void DoesNotDisposeWhenSetToSameValue()
             {
-                var root = new PublicFolder();
-                var folder = GenerateUniqueString();
+                var watcher = Substitute.For<FileSystemWatcher>();
+                var folder = new PublicFolder {Watcher =  watcher};
 
-                var folderpath = Directory.CreateDirectory(Path.Combine(root.FolderPath, folder)).FullName;
-                if (!Directory.Exists(folderpath)) throw new Exception("Folder to test did not get created");
+                folder.Watcher = watcher;
+
+                watcher.DidNotReceive().Dispose();
```

Comment:
`DidNotReceive` is an extension from NSubstitute.

In the setter for the `Watcher` property, the dispose method is called on the existing `FileSystemWatcher` before the new one is assigned. That way we don't have a bunch of file system watchers hanging around after they shouldn't be.

However, if they make the mistake of assigning the same value back to the watcher, without this check I'd end up prematurely disposing of the file system watcher, without them being aware of it.

I'm really just trying to prevent the user from doing something that will cause problems for them that will be difficult for them to troubleshoot. If they call `Dispose` on the watcher themselves, I can't stop that, but it would be easy for them to track that down.